File: list_sorter.py
#!/usr/bin/env python3
"""
list_sorter.py - Сортировщик списков на Python (CLI)
Поддерживает: 6 алгоритмов, визуализацию, замер времени, экспорт.
"""
import sys
import time
import random
import json
import csv
from typing import List, Dict, Callable, Tuple

# ========== АЛГОРИТМЫ СОРТИРОВКИ ==========
def bubble_sort(arr: List) -> List:
    """Пузырьковая сортировка"""
    n = len(arr)
    for i in range(n):
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr

def quick_sort(arr: List) -> List:
    """Быстрая сортировка (рекурсивная)"""
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)

def merge_sort(arr: List) -> List:
    """Сортировка слиянием"""
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)

def merge(left: List, right: List) -> List:
    result = []
    i, j = 0, 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result

def selection_sort(arr: List) -> List:
    """Сортировка выбором"""
    n = len(arr)
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr

def insertion_sort(arr: List) -> List:
    """Сортировка вставками"""
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

def builtin_sort(arr: List) -> List:
    """Стандартная сортировка Python"""
    return sorted(arr)

ALGORITHMS = {
    '1': ('Пузырьковая', bubble_sort),
    '2': ('Быстрая', quick_sort),
    '3': ('Слиянием', merge_sort),
    '4': ('Стандартная', builtin_sort),
    '5': ('Выбором', selection_sort),
    '6': ('Вставками', insertion_sort),
}

# ========== ВСПОМОГАТЕЛЬНЫЕ ФУНКЦИИ ==========
def measure_time(func: Callable, data: List) -> Tuple[List, float]:
    """Измеряет время выполнения сортировки"""
    arr = data.copy()
    start = time.perf_counter()
    result = func(arr)
    elapsed = time.perf_counter() - start
    return result, elapsed

def generate_random_list(size: int, min_val: int = 1, max_val: int = 100) -> List:
    return [random.randint(min_val, max_val) for _ in range(size)]

def print_table(results: Dict[str, Tuple[List, float]], top_n: int = 20):
    """Выводит таблицу результатов с ASCII-гистограммой"""
    if not results:
        print("Нет данных.")
        return
    max_time = max(r[1] for r in results.values())
    print("\n" + "=" * 70)
    print("РЕЗУЛЬТАТЫ СОРТИРОВКИ".center(70))
    print("=" * 70)
    for name, (sorted_data, elapsed) in results.items():
        bar_len = int((elapsed / max_time) * 40) if max_time > 0 else 0
        bar = "█" * bar_len + "░" * (40 - bar_len)
        print(f"{name:<15} {elapsed:.6f} сек.  {bar}")
    print("=" * 70)
    # Показать первые элементы
    first_result = next(iter(results.values()))[0]
    if len(first_result) <= top_n:
        print("Отсортированный список:", first_result)
    else:
        print(f"Отсортированный список (первые {top_n}):", first_result[:top_n])

def export_csv(results: Dict[str, Tuple[List, float]], filename: str):
    with open(filename, 'w', newline='', encoding='utf-8') as f:
        writer = csv.writer(f)
        writer.writerow(['Алгоритм', 'Время (сек)', 'Размер списка'])
        for name, (_, elapsed) in results.items():
            writer.writerow([name, elapsed, len(next(iter(results.values()))[0])])

def export_json(results: Dict[str, Tuple[List, float]], filename: str):
    data = {
        name: {
            'sorted': sorted_data,
            'time': elapsed,
            'size': len(sorted_data)
        }
        for name, (sorted_data, elapsed) in results.items()
    }
    with open(filename, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

def interactive():
    print("📊 СОРТИРОВЩИК СПИСКОВ")
    while True:
        print("\nВыберите действие:")
        print("1. Сортировать введённый список")
        print("2. Сгенерировать случайный список")
        print("3. Сравнить все алгоритмы")
        print("4. Настройки")
        print("0. Выход")
        choice = input("Ваш выбор: ").strip()
        if choice == '0':
            break
        elif choice == '1':
            try:
                data = list(map(int, input("Введите числа через пробел: ").split()))
                if not data:
                    print("Список пуст.")
                    continue
            except ValueError:
                print("Введите числа через пробел.")
                continue
            results = {}
            print("\nВыберите алгоритм (или 'all' для всех):")
            for key, (name, _) in ALGORITHMS.items():
                print(f"{key}. {name}")
            algo_choice = input("Ваш выбор: ").strip()
            if algo_choice == 'all':
                for key, (name, func) in ALGORITHMS.items():
                    sorted_data, elapsed = measure_time(func, data)
                    results[name] = (sorted_data, elapsed)
            elif algo_choice in ALGORITHMS:
                name, func = ALGORITHMS[algo_choice]
                sorted_data, elapsed = measure_time(func, data)
                results[name] = (sorted_data, elapsed)
            else:
                print("Неверный выбор.")
                continue
            print_table(results)
            export = input("Экспортировать результаты? (y/n): ").strip().lower()
            if export == 'y':
                fmt = input("Формат (csv/json): ").strip().lower()
                filename = input("Имя файла: ").strip()
                if not filename:
                    filename = "results." + fmt
                if fmt == 'csv':
                    export_csv(results, filename)
                else:
                    export_json(results, filename)
                print(f"Экспортировано в {filename}")
        elif choice == '2':
            try:
                size = int(input("Размер списка: "))
                min_val = int(input("Минимальное значение: "))
                max_val = int(input("Максимальное значение: "))
                data = generate_random_list(size, min_val, max_val)
                print("Сгенерированный список:", data)
                results = {}
                print("\nВыберите алгоритм (или 'all' для всех):")
                for key, (name, _) in ALGORITHMS.items():
                    print(f"{key}. {name}")
                algo_choice = input("Ваш выбор: ").strip()
                if algo_choice == 'all':
                    for key, (name, func) in ALGORITHMS.items():
                        sorted_data, elapsed = measure_time(func, data)
                        results[name] = (sorted_data, elapsed)
                elif algo_choice in ALGORITHMS:
                    name, func = ALGORITHMS[algo_choice]
                    sorted_data, elapsed = measure_time(func, data)
                    results[name] = (sorted_data, elapsed)
                else:
                    print("Неверный выбор.")
                    continue
                print_table(results)
            except ValueError:
                print("Введите корректные числа.")
        elif choice == '3':
            try:
                data = list(map(int, input("Введите числа через пробел (или оставьте пустым для случайных): ").split()))
                if not data:
                    size = int(input("Размер случайного списка: "))
                    data = generate_random_list(size)
                    print("Сгенерированный список:", data)
                results = {}
                for key, (name, func) in ALGORITHMS.items():
                    sorted_data, elapsed = measure_time(func, data)
                    results[name] = (sorted_data, elapsed)
                print_table(results)
                export = input("Экспортировать результаты? (y/n): ").strip().lower()
                if export == 'y':
                    fmt = input("Формат (csv/json): ").strip().lower()
                    filename = input("Имя файла: ").strip()
                    if not filename:
                        filename = "results." + fmt
                    if fmt == 'csv':
                        export_csv(results, filename)
                    else:
                        export_json(results, filename)
                    print(f"Экспортировано в {filename}")
            except ValueError:
                print("Введите корректные числа.")
        else:
            print("Неверный выбор.")

if __name__ == "__main__":
    if len(sys.argv) > 1:
        # CLI с аргументами
        import argparse
        parser = argparse.ArgumentParser(description="Сортировщик списков")
        parser.add_argument("--list", nargs="+", type=int, help="Список чисел")
        parser.add_argument("--random", type=int, help="Размер случайного списка")
        parser.add_argument("--algorithm", choices=list(ALGORITHMS.keys()) + ['all'], help="Алгоритм")
        parser.add_argument("--export-csv", help="Экспорт в CSV")
        parser.add_argument("--export-json", help="Экспорт в JSON")
        parser.add_argument("--compare", action="store_true", help="Сравнить все алгоритмы")
        args = parser.parse_args()
        data = []
        if args.list:
            data = args.list
        elif args.random:
            data = generate_random_list(args.random)
        else:
            print("Укажите --list или --random")
            sys.exit(1)
        results = {}
        if args.compare or args.algorithm == 'all':
            for key, (name, func) in ALGORITHMS.items():
                sorted_data, elapsed = measure_time(func, data)
                results[name] = (sorted_data, elapsed)
        elif args.algorithm in ALGORITHMS:
            name, func = ALGORITHMS[args.algorithm]
            sorted_data, elapsed = measure_time(func, data)
            results[name] = (sorted_data, elapsed)
        else:
            print("Укажите --algorithm или --compare")
            sys.exit(1)
        print_table(results)
        if args.export_csv:
            export_csv(results, args.export_csv)
            print(f"Экспортировано в {args.export_csv}")
        if args.export_json:
            export_json(results, args.export_json)
            print(f"Экспортировано в {args.export_json}")
    else:
        interactive()

File: list_sorter.js
#!/usr/bin/env node
/**
 * list_sorter.js - Сортировщик списков на JavaScript (Node.js CLI + веб)
 * CLI: node list_sorter.js --list 5 3 8 1 --compare
 * Веб: откройте как HTML
 */
const fs = require('fs');
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// ========== АЛГОРИТМЫ СОРТИРОВКИ ==========
function bubbleSort(arr) {
    const result = [...arr];
    const n = result.length;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n - i - 1; j++) {
            if (result[j] > result[j + 1]) {
                [result[j], result[j + 1]] = [result[j + 1], result[j]];
            }
        }
    }
    return result;
}

function quickSort(arr) {
    if (arr.length <= 1) return arr;
    const pivot = arr[Math.floor(arr.length / 2)];
    const left = arr.filter(x => x < pivot);
    const middle = arr.filter(x => x === pivot);
    const right = arr.filter(x => x > pivot);
    return [...quickSort(left), ...middle, ...quickSort(right)];
}

function mergeSort(arr) {
    if (arr.length <= 1) return arr;
    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));
    return merge(left, right);
}

function merge(left, right) {
    const result = [];
    let i = 0, j = 0;
    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i++]);
        } else {
            result.push(right[j++]);
        }
    }
    return [...result, ...left.slice(i), ...right.slice(j)];
}

function selectionSort(arr) {
    const result = [...arr];
    const n = result.length;
    for (let i = 0; i < n; i++) {
        let minIdx = i;
        for (let j = i + 1; j < n; j++) {
            if (result[j] < result[minIdx]) minIdx = j;
        }
        [result[i], result[minIdx]] = [result[minIdx], result[i]];
    }
    return result;
}

function insertionSort(arr) {
    const result = [...arr];
    for (let i = 1; i < result.length; i++) {
        const key = result[i];
        let j = i - 1;
        while (j >= 0 && result[j] > key) {
            result[j + 1] = result[j];
            j--;
        }
        result[j + 1] = key;
    }
    return result;
}

function builtinSort(arr) {
    return [...arr].sort((a, b) => a - b);
}

const ALGORITHMS = {
    '1': { name: 'Пузырьковая', func: bubbleSort },
    '2': { name: 'Быстрая', func: quickSort },
    '3': { name: 'Слиянием', func: mergeSort },
    '4': { name: 'Стандартная', func: builtinSort },
    '5': { name: 'Выбором', func: selectionSort },
    '6': { name: 'Вставками', func: insertionSort },
};

function measureTime(func, data) {
    const arr = [...data];
    const start = performance.now();
    const result = func(arr);
    const elapsed = (performance.now() - start) / 1000;
    return { result, elapsed };
}

function generateRandomList(size, minVal = 1, maxVal = 100) {
    return Array.from({ length: size }, () => Math.floor(Math.random() * (maxVal - minVal + 1)) + minVal);
}

function printTable(results, topN = 20) {
    if (!Object.keys(results).length) {
        console.log('Нет данных.');
        return;
    }
    const maxTime = Math.max(...Object.values(results).map(r => r.elapsed));
    console.log('\n' + '='.repeat(70));
    console.log('РЕЗУЛЬТАТЫ СОРТИРОВКИ'.padStart(35).padEnd(70));
    console.log('='.repeat(70));
    for (const [name, { elapsed }] of Object.entries(results)) {
        const barLen = maxTime > 0 ? Math.floor((elapsed / maxTime) * 40) : 0;
        const bar = '█'.repeat(barLen) + '░'.repeat(40 - barLen);
        console.log(`${name.padEnd(15)} ${elapsed.toFixed(6)} сек.  ${bar}`);
    }
    console.log('='.repeat(70));
    const firstResult = Object.values(results)[0]?.result || [];
    if (firstResult.length <= topN) {
        console.log('Отсортированный список:', firstResult);
    } else {
        console.log(`Отсортированный список (первые ${topN}):`, firstResult.slice(0, topN));
    }
}

function exportCSV(results, filename) {
    const lines = ['Алгоритм,Время (сек),Размер списка'];
    const size = Object.values(results)[0]?.result.length || 0;
    for (const [name, { elapsed }] of Object.entries(results)) {
        lines.push(`${name},${elapsed},${size}`);
    }
    fs.writeFileSync(filename, lines.join('\n'), 'utf8');
}

function exportJSON(results, filename) {
    const data = {};
    for (const [name, { result, elapsed }] of Object.entries(results)) {
        data[name] = { sorted: result, time: elapsed, size: result.length };
    }
    fs.writeFileSync(filename, JSON.stringify(data, null, 2), 'utf8');
}

function prompt(query) {
    return new Promise(resolve => rl.question(query, resolve));
}

async function interactive() {
    console.log('📊 СОРТИРОВЩИК СПИСКОВ');
    while (true) {
        console.log('\nВыберите действие:');
        console.log('1. Сортировать введённый список');
        console.log('2. Сгенерировать случайный список');
        console.log('3. Сравнить все алгоритмы');
        console.log('4. Настройки');
        console.log('0. Выход');
        const choice = await prompt('Ваш выбор: ');
        if (choice === '0') break;
        else if (choice === '1') {
            const input = await prompt('Введите числа через пробел: ');
            const data = input.split(/\s+/).map(Number).filter(n => !isNaN(n));
            if (!data.length) { console.log('Список пуст.'); continue; }
            const results = {};
            console.log('\nВыберите алгоритм (или all для всех):');
            for (const [key, { name }] of Object.entries(ALGORITHMS)) {
                console.log(`${key}. ${name}`);
            }
            const algoChoice = await prompt('Ваш выбор: ');
            if (algoChoice === 'all') {
                for (const [key, { name, func }] of Object.entries(ALGORITHMS)) {
                    const { result, elapsed } = measureTime(func, data);
                    results[name] = { result, elapsed };
                }
            } else if (ALGORITHMS[algoChoice]) {
                const { name, func } = ALGORITHMS[algoChoice];
                const { result, elapsed } = measureTime(func, data);
                results[name] = { result, elapsed };
            } else {
                console.log('Неверный выбор.');
                continue;
            }
            printTable(results);
            const exportChoice = await prompt('Экспортировать результаты? (y/n): ');
            if (exportChoice.toLowerCase() === 'y') {
                const fmt = await prompt('Формат (csv/json): ');
                const filename = await prompt('Имя файла: ') || `results.${fmt}`;
                if (fmt === 'csv') exportCSV(results, filename);
                else exportJSON(results, filename);
                console.log(`Экспортировано в ${filename}`);
            }
        } else if (choice === '2') {
            try {
                const size = parseInt(await prompt('Размер списка: '));
                const minVal = parseInt(await prompt('Минимальное значение: '));
                const maxVal = parseInt(await prompt('Максимальное значение: '));
                const data = generateRandomList(size, minVal, maxVal);
                console.log('Сгенерированный список:', data);
                const results = {};
                console.log('\nВыберите алгоритм (или all для всех):');
                for (const [key, { name }] of Object.entries(ALGORITHMS)) {
                    console.log(`${key}. ${name}`);
                }
                const algoChoice = await prompt('Ваш выбор: ');
                if (algoChoice === 'all') {
                    for (const [key, { name, func }] of Object.entries(ALGORITHMS)) {
                        const { result, elapsed } = measureTime(func, data);
                        results[name] = { result, elapsed };
                    }
                } else if (ALGORITHMS[algoChoice]) {
                    const { name, func } = ALGORITHMS[algoChoice];
                    const { result, elapsed } = measureTime(func, data);
                    results[name] = { result, elapsed };
                } else {
                    console.log('Неверный выбор.');
                    continue;
                }
                printTable(results);
            } catch (e) {
                console.log('Ошибка ввода.');
            }
        } else if (choice === '3') {
            const input = await prompt('Введите числа через пробел (или оставьте пустым для случайных): ');
            let data = input.split(/\s+/).map(Number).filter(n => !isNaN(n));
            if (!data.length) {
                const size = parseInt(await prompt('Размер случайного списка: '));
                data = generateRandomList(size);
                console.log('Сгенерированный список:', data);
            }
            const results = {};
            for (const [key, { name, func }] of Object.entries(ALGORITHMS)) {
                const { result, elapsed } = measureTime(func, data);
                results[name] = { result, elapsed };
            }
            printTable(results);
            const exportChoice = await prompt('Экспортировать результаты? (y/n): ');
            if (exportChoice.toLowerCase() === 'y') {
                const fmt = await prompt('Формат (csv/json): ');
                const filename = await prompt('Имя файла: ') || `results.${fmt}`;
                if (fmt === 'csv') exportCSV(results, filename);
                else exportJSON(results, filename);
                console.log(`Экспортировано в ${filename}`);
            }
        } else {
            console.log('Неверный выбор.');
        }
    }
    rl.close();
}

function main() {
    const args = process.argv.slice(2);
    if (args.length > 0) {
        const parsed = {};
        for (let i = 0; i < args.length; i++) {
            if (args[i] === '--list') {
                parsed.list = [];
                while (i + 1 < args.length && !args[i+1].startsWith('--')) {
                    parsed.list.push(parseInt(args[++i]));
                }
            } else if (args[i] === '--random') parsed.random = parseInt(args[++i]);
            else if (args[i] === '--algorithm') parsed.algorithm = args[++i];
            else if (args[i] === '--export-csv') parsed.exportCsv = args[++i];
            else if (args[i] === '--export-json') parsed.exportJson = args[++i];
            else if (args[i] === '--compare') parsed.compare = true;
        }
        let data = [];
        if (parsed.list && parsed.list.length) data = parsed.list;
        else if (parsed.random) data = generateRandomList(parsed.random);
        else { console.log('Укажите --list или --random'); process.exit(1); }
        const results = {};
        if (parsed.compare || parsed.algorithm === 'all') {
            for (const [key, { name, func }] of Object.entries(ALGORITHMS)) {
                const { result, elapsed } = measureTime(func, data);
                results[name] = { result, elapsed };
            }
        } else if (ALGORITHMS[parsed.algorithm]) {
            const { name, func } = ALGORITHMS[parsed.algorithm];
            const { result, elapsed } = measureTime(func, data);
            results[name] = { result, elapsed };
        } else {
            console.log('Укажите --algorithm или --compare');
            process.exit(1);
        }
        printTable(results);
        if (parsed.exportCsv) { exportCSV(results, parsed.exportCsv); console.log(`Экспортировано в ${parsed.exportCsv}`); }
        if (parsed.exportJson) { exportJSON(results, parsed.exportJson); console.log(`Экспортировано в ${parsed.exportJson}`); }
    } else {
        interactive().catch(console.error);
    }
}

if (require.main === module) {
    main();
}

// ========== Браузерная версия ==========
if (typeof window !== 'undefined') {
    window.ALGORITHMS = ALGORITHMS;
    window.measureTime = measureTime;
    window.generateRandomList = generateRandomList;
    window.printTable = printTable;
    window.exportCSV = exportCSV;
    window.exportJSON = exportJSON;
}
